app/uk/gov/hmrc/apipublisher/controllers/PublisherController.scala

```diff
@@ -55,17 +56,45 @@ class PublisherController @Inject()(publisherService: PublisherService,
 
   private def publishService(serviceLocation: ServiceLocation)(implicit hc: HeaderCarrier): Future[Result] = {
     Logger.info(s"Publishing service $serviceLocation")
-    publisherService.publishAPIDefinitionAndScopes(serviceLocation).map {
-      case Some(true) =>
-        Logger.info(s"Successfully published API Definition and Scopes for ${serviceLocation.serviceName}")
-        NoContent
-      case Some(false) =>
-        Logger.info(s"Publication awaiting approval for ${serviceLocation.serviceName}")
-        Accepted
-      case None =>
-        Logger.info(s"API publishing disabled for ${serviceLocation.serviceName}")
-        NoContent
-    } recover recovery(s"$FAILED_TO_PUBLISH ${serviceLocation.serviceName}")
+
+    import cats.data.{OptionT, EitherT}
+    import cats.implicits._
+
+    type Over[A] = EitherT[Future,Result, A]
+
+    def getDefinition: Over[ApiAndScopes] = {
+      EitherT(definitionService.getDefinition(serviceLocation).map(_.toRight(BadRequest)))
+    }
+
+    def validate(apiAndScopes: ApiAndScopes): Over[Unit] = {
+      EitherT.fromOptionF(
+        OptionT(publisherService.validateAPIDefinitionAndScopes(apiAndScopes)).map { e =>
+          Console.println("***** "+e)
+          BadRequest(e)
+        }.value
+        , ()
+      ).swap
+    }
+
+    def publish(apiAndScopes: ApiAndScopes): Future[Result] = {
+      publisherService.publishAPIDefinitionAndScopes(serviceLocation, apiAndScopes).map {
+        case Some(true) =>
+          Logger.info(s"Successfully published API Definition and Scopes for ${serviceLocation.serviceName}")
+          NoContent
+        case Some(false) =>
+          Logger.info(s"Publication awaiting approval for ${serviceLocation.serviceName}")
+          Accepted
+        case None =>
```

Comment:
I think this branch is unreachable now. It used to be none if the service didn't have an api definition but we're not calling that endpoint from the publisher service anymore (it used to be in line 70 of the publisher service in the old code)

Reply:
If that's true, we may also want to turn the option of boolean into just a boolean

Reply:
Agreed.  Will trim the types.